OrbitService/LinuxTracingGrpcHandler.cpp

```diff
@@ -163,31 +163,42 @@ uint64_t LinuxTracingGrpcHandler::InternStringIfNecessaryAndGetKey(
 }
 
 void LinuxTracingGrpcHandler::SenderThread() {
-  constexpr std::chrono::duration kSendEvery = std::chrono::milliseconds{20};
-  std::chrono::time_point last_sent = std::chrono::steady_clock::now();
+  constexpr absl::Duration kSendTimeInterval = absl::Milliseconds(20);
+  // This should be lower than kMaxEventsPerResponse in SendBufferedEvents as
+  // a few more events are likely to arrive after the condition becomes true.
+  constexpr uint64_t kSendEventInterval = 5000;
+
   while (tracer_ != nullptr) {
-    std::chrono::duration since_sent =
-        std::chrono::steady_clock::now() - last_sent;
-    if (since_sent < kSendEvery) {
-      std::this_thread::sleep_for(kSendEvery - since_sent);
-    }
-    last_sent = std::chrono::steady_clock::now();
-    SendBufferedEvents();
+    event_buffer_mutex_.LockWhenWithTimeout(
+        absl::Condition(
+            +[](std::vector<CaptureEvent>* event_buffer) {
+              return event_buffer->size() >= kSendEventInterval;
```

Comment:
maybe also check for exit condition here? in which case mutex should also guard tracer_. In this case we won't be stuck for 20ms after everything is already done.

Reply:
As I've said in the past I'm not really sure how all this condition-variable-without-notify works, but `tracer_` is not part of the state protected by the mutex, so it shouldn't be used in the condition. Maybe it's fine if it's not part of the state in this case? Maybe it should be part of such state?

Reply:
Ah sorry, you had already said that the mutex should guard `tracer_`, so I'll do that.

Reply:
Ok, done, please have a look as this is getting tricky.

Reply:
lgtm